driver/labjack/writer_sink.cpp

```diff
@@ -0,0 +1,277 @@
+    if (data_type == synnax::INT16)
+        return synnax::Series(static_cast<int16_t>(val), synnax::INT16);
+    if (data_type == synnax::INT8)
+        return synnax::Series(static_cast<int8_t>(val), synnax::INT8);
+    if (data_type == synnax::UINT32)
+        return synnax::Series(static_cast<uint32_t>(val), synnax::UINT32);
+    if (data_type == synnax::SY_UINT16)
+        return synnax::Series(static_cast<uint16_t>(val), synnax::SY_UINT16);
+    if (data_type == synnax::SY_UINT8)
+        return synnax::Series(static_cast<uint8_t>(val), synnax::SY_UINT8);
+    LOG(ERROR) << "[labjack.writer] Invalid data type";
+}
+
+double series_to_val(const synnax::Series &series) {
+    synnax::DataType data_type = series.data_type;
+    if (data_type == synnax::FLOAT64)
+        return static_cast<double>(series.values<double>()[0]);
+    if (data_type == synnax::FLOAT32)
+        return static_cast<double>(series.values<float>()[0]);
+    if (data_type == synnax::INT32)
+        return static_cast<double>(series.values<int32_t>()[0]);
+    if (data_type == synnax::INT16)
+        return static_cast<double>(series.values<int16_t>()[0]);
+    if (data_type == synnax::INT8)
+        return static_cast<double>(series.values<int8_t>()[0]);
+    if (data_type == synnax::UINT32)
+        return static_cast<double>(series.values<uint32_t>()[0]);
+    if (data_type == synnax::SY_UINT16)
+        return static_cast<double>(series.values<uint16_t>()[0]);
+    if (data_type == synnax::SY_UINT8)
+        return static_cast<double>(series.values<uint8_t>()[0]);
+    LOG(ERROR) << "[labjack.writer] Invalid data type";
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                    StateSource                                //
+///////////////////////////////////////////////////////////////////////////////////
+labjack::StateSource::StateSource(
+    const synnax::Rate state_rate,
+    std::vector<synnax::ChannelKey> state_index_keys,
+    const std::map<synnax::ChannelKey, out_state> state_map
+) : state_rate(state_rate),
+    state_map(state_map),
+    state_index_keys(state_index_keys) {
+    this->timer = loop::Timer(this->state_rate);
+}
+
+
+std::pair<synnax::Frame, freighter::Error> labjack::StateSource::read(
+    breaker::Breaker &breaker) {
+    std::unique_lock<std::mutex> lock(this->state_mutex);
+    this->timer.wait(breaker); // sleep for state period
+    waiting_reader.wait_for(lock, this->state_rate.period().chrono());
+    return std::make_pair(this->get_state(), freighter::NIL);
+}
+
+synnax::Frame labjack::StateSource::get_state() {
+    // frame size = # monitored states + # index channels for the states
+    auto state_frame = synnax::Frame(this->state_map.size() + this->state_index_keys.size());
+
+    for (auto key: this->state_index_keys) {
+        state_frame.add(
+            key,
+            synnax::Series(
+                synnax::TimeStamp::now().value,
```

Comment:
Instantiate the timestamp once so all series are guaranteed to have the same timestamp value.